File: internal/pkg/pinyin/lex_test.go
package pinyin

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestLex(t *testing.T) {
	t.Parallel()

	assertLex(t, "", "")
	assertLex(t, "shi4", "shi-4")
	assertLex(t, "shi4de5", "shi-4 de-5")
	assertLex(t, "shi4de5", "shi-4 de-5")
	assertLex(t, "dou1/Du1/du1", "dou-1", "Du-1", "du-1")
	assertLex(t, "dou1/Du1/du1", "dou-1", "Du-1", "du-1")
	assertLex(t, "yi1 kong3 zhi1 jian4", "yi-1 kong-3 zhi-1 jian-4")
	assertLex(t, "xu1yao4/xiang3", "xu-1 yao-4", "xiang-3")
	assertLex(t, "xu1 yao4/xiang3", "xu-1 yao-4", "xiang-3")
	assertLex(t, " xu1 yao4 / xiang3 ", "xu-1 yao-4", "xiang-3")

	assertLex(t, "jiang14qiang1", "jiang-14 qiang-1")
	assertLex(t, "jiang14/qiang1", "jiang-14", "qiang-1")
}

func assertLex(t *testing.T, src string, expected ...string) bool {
	t.Helper()

	tokenses, err := Lex(src)
	var actual []string
	for _, tokens := range tokenses {
		var chars []string
		for _, token := range tokens {
			chars = append(chars, token.String())
		}
		actual = append(actual, strings.Join(chars, " "))
	}
	require.NoError(t, err)
	return assert.Equal(t, expected, actual)
}

File: electron/src/pinyin/Lex.test.js
import Lex from './Lex';

test('Test Lex', () => {
	assertLex("", "");
	assertLex("shi4", "shi-4");
	assertLex("shi4de5", "shi-4 de-5");
	assertLex("shi4de5", "shi-4 de-5");
	assertLex("dou1/Du1/du1", "dou-1", "Du-1", "du-1");
	assertLex("dou1/Du1/du1", "dou-1", "Du-1", "du-1");
	assertLex("yi1 kong3 zhi1 jian4", "yi-1 kong-3 zhi-1 jian-4");
	assertLex("xu1yao4/xiang3", "xu-1 yao-4", "xiang-3");
	assertLex("xu1 yao4/xiang3", "xu-1 yao-4", "xiang-3");
	assertLex(" xu1 yao4 / xiang3 ", "xu-1 yao-4", "xiang-3");

	assertLex("jiang14qiang1", "jiang-14 qiang-1");
	assertLex("jiang14/qiang1", "jiang-14", "qiang-1");
});

function assertLex(src: string, ...expected: string[]): boolean {
	const tokenses = Lex(src);
	let actual: string[] = [];
	for (const tokens of tokenses) {
		let chars: string[] = [];
		for (const token of tokens.tokens) {
			chars.push(token.string);
		}
		actual.push(chars.join(' '));
	}
	return expected === actual;
}
